File: src/main/java/com/somefrills/features/farming/autofarmer/DiagonalMovement.java
package com.somefrills.features.farming.autofarmer;

/**
 * Diagonal pattern farming (sugar cane, sunflower, moonflower, rose).
 * Cycles through: Forward+Left → Backward → Forward+Left
 */
public class DiagonalMovement implements MovementStrategy {
    private State currentState = State.FORWARD_LEFT;

    public DiagonalMovement() {
    }

    public MovementState getCurrentState() {
        return currentState.toMovementState();
    }

    public void nextState() {
        currentState = currentState.next();
    }

    private enum State {
        FORWARD_LEFT {
            @Override
            public State next() {
                return BACKWARD;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, true, false);
            }
        },
        BACKWARD {
            @Override
            public State next() {
                return FORWARD_LEFT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(false, true, false, false);
            }
        };

        protected abstract State next();

        protected abstract MovementState toMovementState();
    }
}

File: src/main/kotlin/com/somefrills/features/farming/autofarmer/DiagonalMovement.kt
package com.somefrills.features.farming.autofarmer

/**
 * Diagonal pattern farming (sugar cane, sunflower, moonflower, rose).
 * Cycles through: Backward+Left → Forward → repeat
 */
class DiagonalMovement : MovementStrategy() {

    private var state: FarmingState = FarmingState.BACKWARD

    override fun getState(): MovementState {
        return state.toMovementState()
    }

    override fun nextState() {
        state = state.next() as FarmingState
    }

    private enum class FarmingState : State {

        BACKWARD {
            override fun next() = LEFT

            override fun toMovementState(): MovementState {
                return MovementState(
                    MovementState.BACKWARD or MovementState.LEFT
                )
            }
        },

        LEFT {
            override fun next() = BACKWARD

            override fun toMovementState(): MovementState {
                return MovementState(MovementState.FORWARD)
            }
        };
    }
}